lib/ui/painting.dart

```diff
@@ -4302,6 +4302,185 @@ class Canvas extends NativeFieldWrapperClass2 {
+  ///
+  /// An example usage to draw an asset:
+  ///
+  /// ```dart
+  /// class CustomAssetImage extends StatefulWidget {
+  ///   final String asset;
+  ///
+  ///   const CustomAssetImage({required this.asset});
+  ///
+  ///   @override
+  ///   _CustomAssetImageState createState() => _CustomAssetImageState();
+  /// }
+  ///
+  /// class _CustomAssetImageState extends State<CustomAssetImage> {
+  ///   // ImageInfo provides informations to be used on the image rendering
+  ///   ImageInfo? imageInfo;
+  ///   late final VoidCallback _removeListener;
+  ///
+  ///   @override
+  ///   void initState() {
+  ///     super.initState();
+  ///
+  ///     // ExactAssetImage don't care about the widget context
+  ///     final imageProvider = ExactAssetImage(widget.asset);
+  ///     final stream = imageProvider.resolve(ImageConfiguration.empty);
+  ///
+  ///     // A image stream is used because the image can change over time, like when
+  ///     // there are many frames (e.g. GIFs) or the resource image was mutated
+  ///     final listener = ImageStreamListener(_setImage);
+  ///
+  ///     // Each time a new frame is provided, set [imageInfo] so that the painter
+  ///     // is rebuilt
+  ///     stream.addListener(listener);
+  ///
+  ///     _removeListener = () => stream.removeListener(listener);
+  ///   }
+  ///
+  ///   void _setImage(ImageInfo info, _) {
+  ///     setState(() {
+  ///       imageInfo = info;
+  ///     });
+  ///   }
+  ///
+  ///   @override
+  ///   Widget build(BuildContext context) {
+  ///     if (imageInfo?.image == null) {
+  ///       return CircularProgressIndicator();
+  ///     } else {
+  ///       final image = imageInfo!.image;
+  ///       final painter = ImagePainter(image: image);
+  ///
+  ///       return CustomPaint(
+  ///         painter: painter,
+  ///         // For proper layouting, set the CustomPaint size to the image's Size
+  ///         size: Size(
+  ///           image.width.toDouble(),
+  ///           image.height.toDouble(),
+  ///         ),
+  ///       );
+  ///     }
+  ///   }
+  ///
+  ///   @override
+  ///   void dispose() {
+  ///     _removeListener();
+  ///     super.dispose();
+  ///   }
+  /// }
+  ///
+  /// class ImagePainter extends CustomPainter {
+  ///   final ui.Image image;
+  ///
+  ///   const ImagePainter({required this.image});
+  ///
+  ///   @override
+  ///   void paint(Canvas canvas, Size size) =>
+  ///       canvas.drawImage(image, Offset.zero, Paint());
+  ///
+  ///   @override
+  ///   bool shouldRepaint(covariant ImagePainter oldDelegate) =>
+  ///       oldDelegate.image != image;
+  /// }
+  /// ```
+  ///
+  /// {@end-tool}
+  ///
+  /// The previous example used [ExactAssetImage] as the provider for
+  /// simplicity. Thus, it can't handle changes in the current [BuildContext].
+  ///
+  /// To provide an [ImageConfiguration] based on the current [BuildContext]
+  /// consider using [AssetImage] with [createLocalImageConfiguration]:
```

Comment:
consider putting the first of these two paragraphs as part of the description before the first example (at the top of the @tool) and the second paragraph at the top of the second @tool (maybe along with more details there too)